### PR TITLE
Metrics relabel from file

### PR DESCRIPTION
**Pull request by amnonh**
This series adds an option to read the relabel config from file.

Most of Scylla's metrics are reported per-shard, some times they are also reported per scheduling
groups or per tables.  With modern hardware, this can quickly grow to a large number of metrics that 
overload Scylla and the  collecting server. 

One of the main issues around metrics reduction is that many of the metrics are only
helpful in certain situations.

For example, Scylla monitoring only looks at a subset of the metrics. So in large deployments
it would be helpful to scrap only those.

An option to do that, would be to mark all dashboards related metrics with a label value, and then Prometheus
will request only metrics with that label value.

There are two main limitations to scrap by label values:
1. some of the metrics we want to report are in seastar, so we'll need to label them somehow (we cannot just add random labels to seastar metrics)
2. things change, new metrics are introduce and we may want them, it's not practicall to re-compile and wait
for a new release whenever we want to change a label just for monitoring.

It will be best to have the option to add metrics freely and choose at runtime what to report. 

This series make use of Seastar API to perform metrics manipulation dynamically. It includes adding, removing, and changing labels and also enable and disable metrics, and enable and disable the skip_when_empty option.

After this series the configuration could be used with:
```--relabel-config-file conf.yaml```

The general logic and format follows Prometheus metrics_relabel_config configuration.

Where the configuration file looks like:
```
$ cat conf.yaml 
relabel_configs:
  - source_labels: [shard]
    action: drop 
    target_label: shard 
    regex: (2)
  - source_labels: [shard]
    action: replace
    target_label: level 
    replacement: $1
    regex: (.*3)

```

### CODE REVIEW

File: db/config.cc [hunk 1]
```diff
@@ -923,10 +926,12 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , wasm_udf_yield_fuel(this, "wasm_udf_yield_fuel", value_status::Used, 100000, "Wasmtime fuel a WASM UDF can consume before yielding")
     , wasm_udf_total_fuel(this, "wasm_udf_total_fuel", value_status::Used, 100000000, "Wasmtime fuel a WASM UDF can consume before termination")
     , wasm_udf_memory_limit(this, "wasm_udf_memory_limit", value_status::Used, 2*1024*1024, "How much memory each WASM UDF can allocate at most")
+    , relabel_file_name(this, "relabel_file_name", value_status::Used, "", "Optionally, read reable config from file")
     , default_log_level(this, "default_log_level", value_status::Used)
     , logger_log_level(this, "logger_log_level", value_status::Used)
     , log_to_stdout(this, "log_to_stdout", value_status::Used)
     , log_to_syslog(this, "log_to_syslog", value_status::Used)
+
```
Comment (denesb) [inline db/config.cc:934]: Unnecessary change

File: db/config.cc [hunk 2]
```diff
@@ -923,10 +926,12 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , wasm_udf_yield_fuel(this, "wasm_udf_yield_fuel", value_status::Used, 100000, "Wasmtime fuel a WASM UDF can consume before yielding")
     , wasm_udf_total_fuel(this, "wasm_udf_total_fuel", value_status::Used, 100000000, "Wasmtime fuel a WASM UDF can consume before termination")
     , wasm_udf_memory_limit(this, "wasm_udf_memory_limit", value_status::Used, 2*1024*1024, "How much memory each WASM UDF can allocate at most")
+    , relabel_file_name(this, "relabel_file_name", value_status::Used, "", "Optionally, read reable config from file")
```
Comment (denesb) [inline db/config.cc:929]: `relabel_config_file` would be a better name

Comment (denesb) [reply db/config.cc:929]: typo: `reable`

File: db/config.cc [hunk 3]
```diff
@@ -1165,4 +1170,74 @@ future<gms::inet_address> resolve(const config_file::named_value<sstring>& addre
     co_return coroutine::exception(std::move(ex));
 }
 
+static future<std::vector<seastar::metrics::relabel_config>> get_relable_from_file(const std::string& name) {
+    return seastar::open_file_dma(name, open_flags::ro).then([](file f) {
+        return f.size().then([f](size_t s) {
+            return do_with(seastar::make_file_input_stream(f), [s](seastar::input_stream<char>& in) {
+                return in.read_exactly(s).then([](temporary_buffer<char> buf) {
```
Comment (denesb) [inline db/config.cc:1177]: You could use coroutines here, it would greatly improve readability.

File: db/config.cc [hunk 4]
```diff
@@ -1165,4 +1170,74 @@ future<gms::inet_address> resolve(const config_file::named_value<sstring>& addre
...
+                    auto yaml = YAML::Load(sstring(buf.begin(), buf.end()));
+                    std::vector<seastar::metrics::relabel_config> relabels;
+                    const YAML::Node& realbels_config = yaml["relabel_configs"];
+                    relabels.resize(realbels_config.size());
```
Comment (denesb) [inline db/config.cc:1181]: typo: `realbels_config`

File: db/config.cc [hunk 5]
```diff
@@ -1165,4 +1170,74 @@ future<gms::inet_address> resolve(const config_file::named_value<sstring>& addre
...
+                        }
+                        if (element["target_label"]) {
+                            relabels[i].target_label = element["target_label"].as<std::string>();
+                        }
+                        if (element["separator"]) {
+                            relabels[i].separator = element["separator"].as<std::string>();
+                        }
+                        if (element["regex"]) {
+                            relabels[i].expr = element["regex"].as<std::string>();
+                        }
```
Comment (denesb) [inline db/config.cc:1209]: Should we complain about an unrecognized element here?

File: db/config.cc [hunk 6]
```diff
@@ -1165,4 +1170,74 @@ future<gms::inet_address> resolve(const config_file::named_value<sstring>& addre
...
+    if (name.empty()) {
+        return make_ready_future<>();
+    }
+
+    std::vector<seastar::metrics::relabel_config> relabels;
+    return get_relable_from_file(name).then([](std::vector<seastar::metrics::relabel_config>&& relabels) {
+        return do_with(std::move(relabels), false, [](const std::vector<seastar::metrics::relabel_config>& relabels, bool& failed) {
+            return smp::invoke_on_all([&relabels, &failed] {
+                return metrics::set_relabel_configs(relabels).then([&failed](const metrics::metric_relabeling_result& result) {
+                    if (result.metrics_relabeled_due_to_collision > 0) {
```
Comment (denesb) [inline db/config.cc:1228]: coroutines

**Review by denesb — CHANGES_REQUESTED** [review]
Although this PR is simple enough, after reviewing it I still have no idea what this relabeling is and why we want it. Please add a short explanation and motivation.

File: db/config.cc (hunk 2 above)
Comment (amnonh) [reply db/config.cc:929]: will fix

File: db/config.cc (hunk 4 above)
Comment (amnonh) [reply db/config.cc:1181]: will fix

File: db/config.cc (hunk 5 above)
Comment (amnonh) [reply db/config.cc:1209]: I think you're right, I'll throw on unrecognized element

File: db/config.cc (hunk 6 above)
Comment (amnonh) [reply db/config.cc:1228]: sure

File: db/config.cc (hunk 3 above)
Comment (amnonh) [reply db/config.cc:1177]: sure

**Comment by amnonh** [comment]
@denesb Thanks for the review, I've forced push V2. The main changes:
1. move to coroutines
2. fixed some typos

I add some motiviation to the overall series

**Review by denesb — APPROVED** [review]
Thanks @amnonh it is much more clear now.

**Comment by amnonh** [comment]
@denesb now that the relevent patches are merged in Seastar, tests are pass

**Comment by denesb** [comment]
@amnonh there are merge concflicts, please rebase.

**Comment by amnonh** [comment]
> @amnonh there are merge concflicts, please rebase.

I know it sounds stupid, but, are you sure? I pulled rebase and I don't get any updates, also all checks have passed, what am I missing?

**Comment by denesb** [comment]
The conflict is possibly with next. Let's wait for promotion and rebase after.

**Comment by amnonh** [comment]
@denesb I've fixed the conflict and forced push a new version
